### PR TITLE
Implement RFC 2052: Epoches

### PR DESCRIPTION
**Pull request by Manishearth**
Todo: 

 - Make epoches affect the fingerprint
 - Tests

cc https://github.com/rust-lang/rust/issues/44581

Rust PR: https://github.com/rust-lang/rust/pull/48014

r? @acrichto

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! Could some tests be added as well? Something like:

* A test to ensure it's unstable
* A test for a malformed epoch key
* A test that we pass the right epoch
* A test that we unconditionally pass the epoch

**Comment by withoutboats** [comment]
I think it would be more user-friendly to call the field in the toml `rust`. While we've talked about the idea among contributors in terms of "epochs," I think in terms of marketing these releases we will not emphasize the "epoch" language. Among the broader user base and potential use base, this will just be known as "Rust 2018."

If/when we grow the ability to specify a particular toolchain version (e.g. `rustc-1.25.0`), that should be called `toolchain`.

File: src/cargo/core/features.rs [hunk 1]
```diff
@@ -40,9 +40,41 @@
...
 
 use util::errors::CargoResult;
 
+/// The epoch of the compiler (RFC 2052)
+///
+/// Pronounced e-poach
+#[derive(Clone, Copy, Debug, Hash, PartialOrd, Ord, Eq, PartialEq)]
+pub enum Epoch {
+    /// The 2015 epoch
+    Epoch2015 = 2015,
```
Comment (withoutboats) [inline src/cargo/core/features.rs]: Cute, but isn't the only effect of this to make this enum a u16?

**Comment by Manishearth** [comment]
Yeah, was writing the tests on the flight. Pushing now, will add the other tests you mentioned too.

File: src/cargo/core/features.rs (hunk 1 above)
Comment (Manishearth) [reply src/cargo/core/features.rs]: It's more there for forcing the PartialOrd to be the right order. But it's not necessary for Cargo; I just copied this definition from the Rust PR (where comparing epochs is more important)

**Comment by Manishearth** [comment]
Pushed up the tests.

Note that right now it demands a string key, `epoch = "2015"`. `epoch = 2015` will not work. I'm opting for a string in case in the future we wish to do unstable epoch release candidates via `epoch = "2018-next"` or whatever. Or even `epoch = "2018-second"` if for some reason we do two epochs in a year.  However I'm not sure how to tell Serde to accept integer keys and turn them into strings.


Regarding the outward facing name of the feature, @aturon, thoughts?

(it's unstable so we don't need to decide this immediately, but there's also no rush on landing this anyway)

**Comment by withoutboats** [comment]
We tend to require strings even when integers could suffice; we don't accept a version number of `1` for example, it would have to be `"1"`.

**Comment by Manishearth** [comment]
wfm. Though I think not allowing `2015` would be a bit surprising

**Comment by alexcrichton** [comment]
@Manishearth wanna switch to `rust` and we can land this on nightly?

**Comment by Manishearth** [comment]
Done

File: tests/package.rs [hunk 2]
```diff
@@ -887,3 +887,120 @@ fn package_two_kinds_of_deps() {
...
+
+Caused by:
+  epoches are unstable
+
+Caused by:
+  feature `epoch` is required
+
+consider adding `cargo-features = [\"epoch\"]` to the manifest
+")));
+}
```
Comment (bjorn3) [inline tests/package.rs]: Missing trailing newline

**Pull request merged**